herbert_core/utilities/strings/str_to_iarray.m

```diff
@@ -1,219 +1,301 @@
 end
-        
-% Find positions of tokens:
-%  [Add final position, so that beg(i):beg(i+1) contains a token, including any trailing delimiters
-%  which it turns out that sscanf happily ignores]
-delim=[0,sort([strfind(strtmp,char(9)),strfind(strtmp,char(32)),strfind(strtmp,',')]),length(strtmp)+1];
-beg=[delim(diff(delim)>1)+1,length(strtmp)+1];
 
-ntok=numel(beg)-1;
+% Remove comment lines and trailing comments (first occurence of '%' or '!')
+cout = cellfun(@strip_comment, cout, 'UniformOutput', false);
+ok = ~cellfun(@isempty, cout);
+cout = cout(ok);
+
+% Remove brackets if has form '[...]'
+cout = cellfun(@strip_square_brackets, cout, 'UniformOutput', false);
+ok = ~cellfun(@isempty, cout);
+cout = cout(ok);
 
-% Catch case of no tokens
-if ntok==0
-    x=[]; le_nmax=true; return
+% Concatenate all strings into one (with a leading space to ensure whitespace delimiter)
+if ~isempty(cout)
+    ctmp = cellfun(@(x)([' ', x]), cout, 'UniformOutput', false);
+    strtmp = strcat(ctmp{:});
+else
+    strtmp = ' ';   % ensures output consistent with non-empty but no-integer string
 end
 
-% Parse each token in turn
-xlo=zeros(1,ntok);
-xhi=zeros(1,ntok);
-n=zeros(1,ntok);
+% Find positions of tokens: 
+% - Delimiters are:
+%   - A comma with any number of whitespace characters before and after
+%    (including no white space): \s*,\s*
+%   - One or more whitespace characters: \s+
+% - Tokens run from iend(i)+1:ibeg(i+1)-1, but we must allow for the case of 
+%   tokens at beginning &/or running to end; the construction below means that
+%   the first and/or last tokens are empty, which are handled in the function
+%   str_token_to_iarray
+[ibeg, iend] = regexp (strtmp, '\s*,\s*|\s+');  % delimiters
+itok_beg = [1,iend+1];
+itok_end = [ibeg-1, numel(strtmp)];
+
+% Parse each token in turn, and pick out only those ranges with at least one point
+ntok=numel(itok_beg);
+x1 = NaN(1, ntok);
+dx = NaN(1, ntok);
+n = NaN(1,ntok);
 for i=1:ntok
-    [xlo(i),xhi(i),n(i)]=str_token_to_iarray(strtmp(beg(i):beg(i+1)-1));
-    if n(i)<0
-        error('Check format of string contents')
-    end
+    [x1(i), dx(i), ~, n(i)] = str_token_to_iarray (strtmp(itok_beg(i):itok_end(i)));
 end
+ok = (n>0);
+ntok = sum(ok);
+x1 = x1(ok);
+dx = dx(ok);
+n = n(ok);
 
 % Fill output array
-nend=cumsum(n);
-nbeg=[1,nend(1:end-1)+1];
-if isfinite(nmax)
-    if nend(ntok)>nmax
-        le_nmax=false;
-        ntok=lower_index(nend,nmax);    % find the number of tokens required to contain nmax integers
-        nend(ntok)=nmax;                % update nend(ntok); we will ignore all later entries
-        n(ntok)=nmax-nbeg(ntok)+1;
-        if xhi(ntok)>xlo(ntok)
-            xhi(ntok)=xlo(ntok)+n(ntok)-1;
-        else
-            xhi(ntok)=xlo(ntok)-n(ntok)+1;
-        end
+if isempty(n)
+    % Case of no integers read
+    x = NaN(1,0);   % to be consistent with e.g. 3:2 which has size == [1,0]
+    le_nmax = true; % nmax is guaranteed to be >= 1
+    
+else
```

Comment:
Generally I like to avoid multiple return locations unless I am catching trivial special cases near the start of a function - I find that multiple returns make code logic harder to follow.